### ISSUE TITLE
constexpr Hashing Enum ?

### ISSUE THREAD

**Issue by calebarist** [Neargye/magic_enum#209]
Essentially, a meaningful association between the stringified enum key and the integral value it maps to is desirable.

When communicating an enum across an interface, too often we end up communicating values that mean absolutely nothing.
If they were, instead, the result of passing a string into a constexpr hashing function, then that would add a strong association and more value to the integral value. Such an enum wrapper (or utility function) could be very useful for interfaces.

Though for our purposes, it would be most helpful to also have a C# companion library with similar functionality at the other side of the interface--the hashing function in particular.

Just an idea, let me know what you think!

**Comment by Neargye** [comment]
Could you give examples? I don’t understand a little why needs hash.

**Comment by calebarist** [comment]
If you don't have a hash you have to share both what the enum is for and the random value assigned to it to the team working on the other side of the interface.

If you can compute the value which will be transmitted from the name of the enum, it would simplify that process and avoid the bugs that would occur from one side sending an enum value different to what is expected. And needing to modify both projects at the same time if the values are changed.

It's a nice to have feature, but a good one in that there is a strong association between the enum element name and the value that it is mapped to.

**Comment by Neargye** [comment]
But then there will be a problem if the name of the enum changes.

**Comment by BrukerJWD** [comment]
Couldn't you just send the enum_name as string if you want to have this constant?

**Comment by calebarist** [comment]
That is true, the name of the enum element would have to be the same for both parties, but the name is meaningful and probably easier to not change than some random integer value.

I imagine the process would go like:
-create enum keys
-stringify enum keys with C preprocessor macro or other tool
-constexpr compute the hash code for the stringified enum keys
-associate the computed value back to the enum key

transmitting strings is more overhead and slightly more difficult, a better standard library enum would probably make the process easier, but there would still need to be a shared hashing fn.

It's just an idea, one that I personally think is a good idea.

**Comment by BrukerJWD** [comment]
But the integer value isn't random :)
If you want specific ones that shall not change you can assign them explicitly to make this clear, otherwise they start at 0 and are incremented. That's definied by the C++ standard and definitely not random.
These values only change if you insert a new enum value before another and only have implicit integers.

    enum class {
        A = 0,
        B = 1,
        // C was removed
        D = 3,
    };

Hashing has a certain probablility of collisions, i.e. the keys cannot be distinguished in this case.

**Comment by calebarist** [comment]
one side of the interface may implement only a subset of the other (being one of many clients, for instance), the integer value associated is certainly random because they might not also apply an ordering to the enum keys, which would be even more information to pass between projects

**Comment by BrukerJWD** [comment]
Sharing common headers used for communication between these projects may be a good idea anyway ;)

If not doing so:
I agree that one would not want to communicate a certain order in the enum. That's why you can make their integer values explicit! If explicit, they are required and cannot be removed. On the other hand, the enum gives freedom to change the name of the enum key to something more appropriate for the current use case, while keeping the underlying integer value.

Also, you ignored my comment about hash collisions. Using std::hash is not possible as its implementation is compiler defined, and you certainly don't want to enforce all projects to use the same compiler version, do you? So one would have to implement a custom hash function, as resistent to collisions as possible. This is difficult, and all one can do is to try to reduce the probability. What do you do if two keys generate the same hash? Do you expect a compiler error? Do you want to be able to "seed" the hash with a different number to make it work for this specific enum?

Personally speaking, I would use the underlying value of enums, because that's easily transportable. You can even define the underlying type and shrink it to 1 byte if this helps with the communication.

Anyway, you can implement your hashing idea yourself, using enum_name and applying a tbd hash function, and on the receiving side have a map which you filled with these hashes at compile time by statically iterating over all enum keys and getting their name.

**Comment by calebarist** [comment]
Often the projects are in different programming languages so sharing headers exists in documentation probably.
Yes it would have to detect collisions and issue an error/warning about them.

But ok if it's not something you're interested in, maybe I will play around with it sometime.

**Comment by BrukerJWD** [comment]
Projects in different programming languages requires this hash functionality to be implemented identically several times. That sounds like a lot to communicate (especially those parts that one doesn't think of the first time), even if only once per lanugage.

I cannot speak for @Neargye, though.

**Comment by Neargye** [comment]
It seems to me that you can make the necessary functionality using the api of this library.
For use in a project with different languages, you can also use your existing hash.  It seems to me wrong to add something from other languages into this library, otherwise will need add to Java, Go and others
